1.6/Source/PawnStorages/PawnStorages/CompPawnStorageNutrition.cs

```diff
@@ -185,35 +247,95 @@ public virtual void EmulateScaledPawnAgeTick(Pawn pawn)
 
-    public List<IntVec3> AdjCellsCardinalInBounds =>
-        cachedAdjCellsCardinal ??= GenAdj.CellsAdjacentCardinal(parent)
-            .Where(c => c.InBounds(parent.Map))
-            .ToList();
 
-    public virtual bool TryAbsorbNutritionFromHopper(float nutrition)
+    public virtual bool IsValidNutritionRequest(float nutrition)
     {
-        if (nutrition <= 0) return false;
-        if (!HasEnoughFeedstockInHoppers()) return false;
+        return nutrition > 0 && HasEnoughFeedstockInHoppers();
+    }
 
-        Thing feedInAnyHopper = FindFeedInAnyHopper();
-        if (feedInAnyHopper == null)
+    public virtual float ProcessFeedAbsorption(Thing feedSource, float nutritionToAbsorb)
+    {
+        /*
+         * Processes the absorption of specified nutrition from a given feed source. Determines the amount of nutrition
+         * that can be absorbed from the provided feed source and updates its state accordingly. The remaining
+         * unabsorbed nutrition is returned.
+         */
+        if (feedSource == null)
+            return nutritionToAbsorb;
+
+        float remainingNutrition = nutritionToAbsorb;
+        float nutritionPerUnit = feedSource.GetStatValue(StatDefOf.Nutrition);
+
+        while (remainingNutrition > 0)
         {
-            return false;
-        }
+            int unitsToAbsorb = Mathf.Min(feedSource.stackCount, Mathf.FloorToInt(remainingNutrition / nutritionPerUnit));
+
+            if (unitsToAbsorb <= 0)
+                break;
+
+            float nutritionAbsorbed = unitsToAbsorb * nutritionPerUnit;
+            if (nutritionAbsorbed <= 0)
+                break;
 
-        int count = Mathf.Min(feedInAnyHopper.stackCount, Mathf.CeilToInt(nutrition / feedInAnyHopper.GetStatValue(StatDefOf.Nutrition)));
-        storedNutrition += count * feedInAnyHopper.GetStatValue(StatDefOf.Nutrition);
+            remainingNutrition -= nutritionAbsorbed;
+
+            if (unitsToAbsorb > 1)
+            {
+                Thing thing = feedSource.SplitOff(unitsToAbsorb);
+                thing.Destroy();
+            }
+            else
+            {
+                feedSource.DeSpawn();
+                break;
+            }
+        }
 
-        feedInAnyHopper.SplitOff(count);
+        return remainingNutrition;
+    }
 
+    public virtual bool AddNutritionToStorage(float amount)
+    {
+        if (amount <= 0 || Mathf.Approximately(amount, 0)) return false;
+        StoredNutrition += amount;
         return true;
     }
 
-    public virtual bool ValidFeedstock(ThingDef def) => def.IsNutritionGivingIngestible && def.ingestible.preferability != FoodPreferability.Undefined;
+    public virtual bool TryAbsorbNutritionFromSource(float requestedNutrition)
+    {
+        // Absorb from hoppers
+        if (requestedNutrition <= 0)
+            return false;
+
+        if (!IsValidNutritionRequest(requestedNutrition))
+            return false;
+
+        Thing feedSource = FindFeedInAnyHopper();
```

Comment:
I think this needs to be a loop because if you have 4 hoppers this will pick one at random and only absorb from that single hopper. It probably needs to repeat until feedsource is null instead of just doing it once. Alternatively where it's called but I think here makes sense.

Reply:
It's already looped in `FeedAndRecordWantedAmount`. 
It calls `AbsorbToFeedIfNeeded` in the while condition, which in turn calls `TryAbsorbNutritionFromSource`. 

`AbsorbNutritionTick` also calls it, which runs every `Props.TicksToAbsorbNutrients` and tries to absorb `TargetNutritionLevel - StoredNutrition`. I think I only did it the once there, as that's how normal hoppers "appear" to work to a player. A single thing's worth of nutrition absorbed at a time every so often until storage is full. 

If we were to put a loop here, the whole stack of things on the hopper could be absorbed to storage all at once. From a technical perspective, same thing just slower, but to a player, might seem weird. 

Will defer to you though on wanted behaviour here. 

I think